File: src/main/java/ee/carlrobert/codegpt/actions/toolwindow/MoveAction.java
package ee.carlrobert.codegpt.actions.toolwindow;

import com.intellij.openapi.actionSystem.ActionUpdateThread;
import com.intellij.openapi.actionSystem.AnAction;
import com.intellij.openapi.actionSystem.AnActionEvent;
import com.intellij.openapi.project.Project;
import ee.carlrobert.codegpt.conversations.Conversation;
import ee.carlrobert.codegpt.conversations.ConversationsState;
import java.util.Optional;
import javax.swing.Icon;
import org.jetbrains.annotations.NotNull;

public abstract class MoveAction extends AnAction {

  private final Runnable onRefresh;

  protected abstract Optional<Conversation> getConversation(@NotNull Project project);

  protected MoveAction(String text, String description, Icon icon, Runnable onRefresh) {
    super(text, description, icon);
    this.onRefresh = onRefresh;
  }

  @Override
  public void update(@NotNull AnActionEvent event) {
    event.getPresentation().setEnabled(ConversationsState.getCurrentConversation() != null);
  }

  @Override
  public void actionPerformed(@NotNull AnActionEvent event) {
    var project = event.getProject();
    if (project != null) {
      getConversation(project)
          .ifPresent(conversation -> {
            ConversationsState.getInstance().setCurrentConversation(conversation);
            onRefresh.run();
          });
    }
  }

  @Override
  public @NotNull ActionUpdateThread getActionUpdateThread() {
    return ActionUpdateThread.BGT;
  }
}

File: src/main/kotlin/ee/carlrobert/codegpt/actions/toolwindow/MoveAction.kt
package ee.carlrobert.codegpt.actions.toolwindow

import com.intellij.openapi.actionSystem.ActionUpdateThread
import com.intellij.openapi.actionSystem.AnAction
import com.intellij.openapi.actionSystem.AnActionEvent
import com.intellij.openapi.project.Project
import ee.carlrobert.codegpt.conversations.Conversation
import ee.carlrobert.codegpt.conversations.ConversationsState
import java.util.Optional
import javax.swing.Icon

abstract class MoveAction protected constructor(
  text: String?, description: String?, icon: Icon?, private val onRefresh: Runnable) : AnAction(text, description, icon) {

  protected abstract fun getConversation(project: Project): Optional<Conversation?>

  override fun update(event: AnActionEvent) {
    event.presentation.isEnabled = ConversationsState.getCurrentConversation() != null
  }

  override fun actionPerformed(event: AnActionEvent) {
    val project = event.project ?: return
    getConversation(project).ifPresent {
      ConversationsState.getInstance().setCurrentConversation(it)
      onRefresh.run()
    }
  }

  override fun getActionUpdateThread(): ActionUpdateThread {
    return ActionUpdateThread.BGT
  }
}
